mypyc/codegen/emitmodule.py

```diff
@@ -918,7 +920,13 @@ def generate_module_def(self, emitter: Emitter, module_name: str, module: Module
 
         emitter.emit_line('return {};'.format(module_static))
         emitter.emit_lines('fail:',
-                           '{} = NULL;'.format(module_static),
+                           'Py_XDECREF({});'.format(module_static),
+                           'Py_XDECREF(modname);')
+        # the type objects returned from CPyType_FromTemplate are all new references
+        # so we have to decref them
+        for t in type_structs:
```

Comment:
We should also do the finals, too. Those need to be reset to whatever the proper error value for the type is.

Reply:
I think the right thing to do is to call out to `emit_dec_ref`, which knows how to decref everything: https://github.com/python/mypy/blob/master/mypyc/codegen/emit.py#L344

And then assign the result of `c_undefined_value` to the variable, to clear it out

Reply:
Yeah, that does look a lot cleaner. This should be fixed in ebe20ed7d03802584fdd593ac77224e9be3f8022.